### PR TITLE
Improve RootSolvers documentations and tests; fix some edge cases

### LINKED ISSUE #21: Add more difficult tests
Right now we’re testing on a very simple problem and it would be nice to add more tests, and more complex/difficult tests to pass. Some good properties might include problems with

 - [ ] asymptotic solutions
 - [ ] multiple/non-unique solutions
 - [ ] solutions at discontinuities

### LINKED ISSUE #47: Docs example should illustrate how to extract the solution value
The example shows how to obtain a `sol::CompactSolution`, but not how to extract the value of the solution (eg a number), which is what we'd typically need to use `find_zero`

https://clima.github.io/RootSolvers.jl/dev/#Usage

### PR DESCRIPTION
**Pull request by tapios**
* Improved robustness by adding catch for zero derivatives in Newton's method and then branching to secant method and adding damping (backtracking line search and step size limits) to Newton's method to prevent overflow
* Added Brent method to `RootSolvers.jl` and test as a robust and locally quadratic bracketing method
* Added more challenging tests (exponentials, multiple roots, high root multiplicity). Closes #21 
* Removed code duplication in `find_zero` with helper functions
* Refactored test suite to be more comprehensive and robust
* Improved and added documentation throughout. Closes #47

### CODE REVIEW

**Comment by tapios** [comment]
@dennisYatunin Please take a look primarily at src/RootSolvers.jl and let me know if you see patterns that should be changed right now. I'll take care of remaining test issues etc.

File: docs/src/GettingStarted.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+**Example: 1 Million problems on the GPU**
+```julia
+using RootSolvers, CUDA
+
+# Create GPU arrays for batch processing
+x0 = CUDA.fill(1.0f0, 1000, 1000)  # 1M initial guesses on GPU
+x1 = CUDA.fill(2.0f0, 1000, 1000)  # Second initial guesses
+
+# Define GPU-compatible function
+f(x) = x.^3 .- x .- 2.0
```
Comment (dennisYatunin) [inline docs/src/GettingStarted.md:176]:
```suggestion
f(x) = x^3 - x - 2
```
Why is broadcasting needed here? Isn't this function called in parallel at every point of the solution array, rather than once for the entire array?

Also, `- 2.0` introduces a `Float32 -> Float64` conversion. Use `- 2` to avoid this.

File: src/RootSolvers.jl [hunk 2]
```diff
@@ -273,35 +544,415 @@ on ``|(x2-x1)/x1| || |x2-x1|``
...
+        b_new_swap = ifelse(cond_swap, a, b)
+        fa_new_swap = ifelse(cond_swap, fb, fa)
+        fb_new_swap = ifelse(cond_swap, fa, fb)
+        a, fa, b, fb = a_new_swap, fa_new_swap, b_new_swap, fb_new_swap
+    end
+
+    return SolutionResults(soltype, b, false, fb, maxiters, x_history, y_history)
+end
+
+# Dedicated helper for the Secant method (an open, two-point method)
```
Comment (haakon-e) [inline src/RootSolvers.jl:826]: I would change this to a proper docstring and add it to developer docs.

File: src/RootSolvers.jl [hunk 3]
```diff
@@ -273,35 +544,415 @@ on ``|(x2-x1)/x1| || |x2-x1|``
...
+        if tol(x0, x1, y1) 
+            return SolutionResults(soltype, x1, true, y1, i, x_history, y_history)
+        end
+    end
+    
+    return SolutionResults(soltype, x1, false, y1, maxiters, x_history, y_history)
+end
+
+
+# Helper: Main iteration loop for Newton methods
```
Comment (haakon-e) [inline src/RootSolvers.jl:874]: I would change this to a proper docstring and add it to developer docs.

File: src/RootSolvers.jl [hunk 4]
```diff
@@ -491,45 +1041,10 @@ function find_zero(
...
-    return SolutionResults(
-        soltype,
-        x0,
-        false,
-        y,
-        maxiters,
-        x_history,
-        y_history,
-    )
+    return _find_zero_newton((x) -> value_deriv(f, x), f, x0, soltype, tol, maxiters)
```
Comment (haakon-e) [inline src/RootSolvers.jl:1044]: I don't _think_ this should matter too much, but I would be inclined to write this as
```suggestion
   return _find_zero_newton(Base.Fix1(value_deriv, f), f, x0, soltype, tol, maxiters)
```
which might help with type stability in some settings.

File: src/RootSolvers.jl [hunk 5]
```diff
@@ -538,40 +1053,57 @@ function find_zero(
...
+
+## Returns
+- `Tuple`: Initial guess(es) specific to the method type
+
+## Example
+```julia
+method = SecantMethod{Float64}(0.0, 1.0)
+args = method_args(method)  # Returns (0.0, 1.0)
+```
+"""
```
Comment (haakon-e) [inline src/RootSolvers.jl:1078]: This and other methods that are in the "Internal methods" section of the API docs would probably be useful to include in/move to a developer docs section.

File: test/runtests.jl [hunk 6]
```diff
@@ -10,81 +11,201 @@ using Test
...
+end
+
+# Helper function to get maxiters based on problem name
+function get_maxiters(problem_name)
+    if problem_name in ["high-multiplicity root", "steep exponential function", "trigonometric function"]
+        return 10_000
+    else
+        return 10_000
+    end
+end
```
Comment (haakon-e) [inline test/runtests.jl:92]: is it intentional to have the same iters in every case?

File: test/runtests.jl
Comment (haakon-e) [inline test/runtests.jl:1]: it would be useful if it was described somewhere (developer docs) how to test the package, and also how to build the docs locally.

File: docs/src/GettingStarted.md [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+# Create GPU arrays for batch processing
+x0 = CUDA.fill(1.0f0, 1000, 1000)  # 1M initial guesses on GPU
+x1 = CUDA.fill(2.0f0, 1000, 1000)  # Second initial guesses
+
+# Define GPU-compatible function
+f(x) = x^3 - x - 2
+
+# Solve all problems in parallel using broadcasting
+method = SecantMethod(x0, x1)
+sol = find_zero.(f, method, CompactSolution())
```
Comment (imreddyTeja) [inline docs/src/GettingStarted.md:177]: I'm not actually sure this example would work. I expect this would throw an isbits error. This is something
that could be supported easilty though

**Comment by tapios** [comment]
> This was a big PR, so the review is correspondingly large!
> 
> Many comments are duplicate to keep track of similar comments/feedback.
> 
> Overall, great work refreshing the package and getting it in great shape! Comments mostly provide suggestions for improved documentation, both for users and potential developers or contributors. If you have any questions about specific feedback, please let me know.
> 
> Assuming these are addressed, I'm happy to see this merged.

Awesome! Thanks so much, Haakon. I'll work through them.

**Review by imreddyTeja — COMMENTED** [review]
I'm not very familiar with KernelAbstractions.jl, but I checked, and everything is type stable. However, the customized broadcasting behavior currently will allocate. I believe this is ok if kernels are launched with kernel abstractions, as that avoids broadcasting. However, this prevents launching kernels by broadcasting on CuArrays (which also currently does not work because methods are not isbits).

File: test/runtests.jl [hunk 8]
```diff
@@ -10,81 +11,201 @@ using Test
...
+
+# Helper function to run solver and extract results
+function run_solver_test(f, method, sol_type, tol, maxiters, is_array)
+    if is_array
+        sol = RootSolvers.find_zero.(Ref(f), method, sol_type, tol, maxiters)
```
Comment (imreddyTeja) [inline test/runtests.jl:97]:
```suggestion
        sol = RootSolvers.find_zero.(f, method, sol_type, tol, maxiters)
```

File: src/RootSolvers.jl [hunk 9]
```diff
@@ -273,35 +544,415 @@ on ``|(x2-x1)/x1| || |x2-x1|``
...
+        y_history = init_history(soltype, y0)
+        return SolutionResults(soltype, x0, false, y0, 0, x_history, y_history)
+    end
+    
+    x_history = init_history(soltype, x0)
+    y_history = init_history(soltype, y0)
+   
+    lastside = 0
+    
+    local x, y
```
Comment (imreddyTeja) [inline src/RootSolvers.jl:702]: I'm not sure if this decleration is needed

**Comment by tapios** [comment]
> I'm not very familiar with KernelAbstractions.jl, but I checked, and everything is type stable. However, the customized broadcasting behavior currently will allocate. I believe this is ok if kernels are launched with kernel abstractions, as that avoids broadcasting. However, this prevents launching kernels by broadcasting on CuArrays (which also currently does not work because methods are not isbits).

Thank you, @imreddyTeja! Do you know how to improve this?

File: docs/src/GettingStarted.md [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+# Find the root of x^2 - 100^2 using the secant method
+sol = find_zero(x -> x^2 - 100^2, SecantMethod(0.0, 1000.0))
+
+if sol.converged
+    println("Root found: ", sol.root)
+else
+    println("Root not found")
+end
+```
```
[the remark this answers is not in the document]
Comment (tapios) [reply docs/src/GettingStarted.md:31]: Done

File: docs/src/GettingStarted.md (hunk 7 above)
Comment (tapios) [reply docs/src/GettingStarted.md:173]: I'd be grateful if you revisit this in a subsequent PR.

File: test/runtests.jl (hunk 6 above)
Comment (tapios) [reply test/runtests.jl:92]: Leftover from experimentation. I removed the if ... else.

File: src/RootSolvers.jl (hunk 3 above)
Comment (tapios) [reply src/RootSolvers.jl:898]: I prefer to leave it so we do not duplicate information in the docs.

File: src/RootSolvers.jl (hunk 5 above)
Comment (tapios) [reply src/RootSolvers.jl:1106]: Done.

File: test/runtests.jl
Comment (tapios) [reply test/runtests.jl:1]: Added Developer Docs.

File: src/RootSolvers.jl [hunk 11]
```diff
@@ -273,35 +544,415 @@ on ``|(x2-x1)/x1| || |x2-x1|``
...
+tol = RelativeOrAbsoluteSolutionTolerance(1e-12, 1e-15)
+sol = find_zero(x -> cos(x), 
+               NewtonsMethodAD{Float64}(1.0), 
+               CompactSolution(), 
+               tol)
+println("π/2 ≈ \$(sol.root)")
+
+# Robust bracketing method for difficult functions
+sol = find_zero(x -> x^3 - 2x - 5, RegulaFalsiMethod{Float64}(2.0, 3.0))
+```
```
[the remark this answers is not in the document]
Comment (tapios) [reply src/RootSolvers.jl:635]: It's a good idea. But I prefer to leave it as it is, in part because I don't want to have to update doc examples when default parameters in the algorithms change.

File: src/RootSolvers.jl [hunk 12]
```diff
@@ -273,35 +544,415 @@ on ``|(x2-x1)/x1| || |x2-x1|``
...
+
+    for i in 1:maxiters
+        # Convergence check
+        if tol(a, b, fb) || fb == 0
+            return SolutionResults(soltype, b, true, fb, i, x_history, y_history)
+        end
+
+        # On GPUs, `ifelse` is often more performant than `if-else` blocks
+        # by avoiding branch divergence.
+        s = ifelse(fa != fc && fb != fc,
```
[the remark this answers is not in the document]
Comment (tapios) [reply src/RootSolvers.jl:804]: Good point. Replaced by bitwise operators
